lib/devise/multi_email/models/confirmable.rb

```diff
@@ -38,64 +38,108 @@ module ConfirmableExtensions
 
         included do
           multi_email_association.include_module(EmailConfirmable)
-        end
-
-        # delegate before creating overriding methods
-        delegate :skip_confirmation!, :skip_confirmation_notification!, :skip_reconfirmation!, :confirmation_required?,
-                 :confirmation_token, :confirmed_at, :confirmation_sent_at, :confirm, :confirmed?, :unconfirmed_email,
-                 :reconfirmation_required?, :pending_reconfirmation?, to: :primary_email_record, allow_nil: true
-
-        # This need to be forwarded to the email that the user logged in with
-        def active_for_authentication?
-          login_email = current_login_email_record
 
-          if login_email && !login_email.primary?
-            super && login_email.active_for_authentication?
-          else
-            super
+          # Handle automatically confirming and linking `unconfirmed_email` if email change is not postponed
+          before_update do
+            if !postponed_email_change? && multi_email.unconfirmed_email_changes?
+              multi_email.set_primary_record_to(
+                multi_email.unconfirmed_email_record,
+                skip_confirmations: true
+              )
+            end
           end
-        end
-
-        # Shows email not confirmed instead of account inactive when the email that user used to login is not confirmed
-        def inactive_message
-          login_email = current_login_email_record
 
-          if login_email && !login_email.primary? && !login_email.confirmed?
-            :unconfirmed
-          else
-            super
+          # Don't send notifications when the email records are saved
+          # when saving the parent record.
+          if respond_to?(:after_commit)
+            after_commit(prepend: true){ multi_email.primary_email_record.try(:skip_confirmation_notification!) }
```

Comment:
@allenwq yeah, the reason is that, when you do things through the parent record, the confirmation notification is sent out. But then ActiveRecord also triggers the confirmation notification on any associated `Email` records separately.

Specifically, the reason this happens is because we use method delegation from parent to `Email` record. The refactor focuses on delegating only the data columns to the `Email` record and all helper methods remain on the parent record (as much as possible). So Devise thinks it's just triggering callbacks on the parent (e.g. `User`) record, but ActiveRecord then triggers callbacks on the `Email` records too. This prevents triggering again on the `Email` records.